src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala

```diff
@@ -0,0 +1,133 @@
+    }
+
+    // Coalesce sequences of values separated by blockSize into single ranges.
+    // The sequence of values must be sorted.
+    def coalesceAdjacent(xs: IndexedSeq[Long], blockSize: Long): Iterator[(Long, Long)] = {
+      val rangeStarts = ArrayBuffer(0)
+      for (i <- 1 until xs.length) {
+        if (xs(i - 1) + blockSize != xs(i)) {
+          rangeStarts += i
+        }
+      }
+      rangeStarts += xs.length
+
+      rangeStarts.sliding(2).map(items => {
+        val startIdx = items(0)
+        val nextStartIdx = items(1)
+        (xs(startIdx), xs(nextStartIdx - 1))
+      })
+    }
+
+    // Returns a sequence of all blocks which overlap the read, its mate or the insert between them.
+    def matedReadBlocks(r: PairedRead[MappedRead]): Seq[Long] = {
+      if (!r.isMateMapped) {
+        Seq[Long]()
+      } else {
+        val mateStart = r.mateAlignmentProperties.map(mp => mp.start).getOrElse(r.read.start)
+        val start = List(r.read.start, r.read.end, mateStart).min
+        val roundedStart = (1.0 * start / BLOCK_SIZE).toLong * BLOCK_SIZE
+        val insertSize = orientedInsertSize(r).getOrElse(0)
+        (roundedStart to (roundedStart + insertSize) by BLOCK_SIZE)
+      }
+    }
+
+    override def run(args: Arguments, sc: SparkContext): Unit = {
+      val readSet = Common.loadReadsFromArguments(args, sc, Read.InputFilters(nonDuplicate = true))
+      val pairedMappedReads = readSet.mappedPairedReads
+      val firstInPair = pairedMappedReads.filter(_.isFirstInPair)
+
+      // Pare down to mate pairs which aren't highly displaced & aren't inverted.
+      val readsInRange = firstInPair.filter(r => {
+        val insertSize = orientedInsertSize(r).getOrElse(0)
+        (insertSize > 0) && (insertSize < MAX_INSERT_SIZE)
+      })
+      readsInRange.persist()
+
+      val insertSizes = readsInRange.flatMap(orientedInsertSize)
+      val insertStats = insertSizes.stats()
+      println("Stats on inferredInsertSize: " + insertStats)
+
+      // Create a map from (block) -> [insert sizes]
+      // TODO: take greater advantage of data locality than .groupByKey() does?
+      val insertsByBlock = readsInRange.flatMap(r => {
+        val insertSize = orientedInsertSize(r).getOrElse(0)
+        matedReadBlocks(r).map(pos => (Location(r.read.referenceContig, pos), insertSize))
+      }).groupByKey()
+
+      // Find blocks where the average insert size is 2 sigma from the average for the whole genome.
+      val exceptionalThreshold = insertStats.mean + 2 * insertStats.stdev
+      val exceptionalBlocks = insertsByBlock.filter(pair => {
+        val inserts = pair._2
+        val stats = new StatCounter()
+        stats.merge(inserts.map(_.toDouble))
+        (stats.mean > exceptionalThreshold)
+      })
+
+      // Group adjacent blocks on the same contig
+      val exceptionalRanges = exceptionalBlocks.groupBy(_._1.contig).flatMap(kv => {
```

Comment:
I am guessing this will work for you with the size the array you anticipate dealing with, but the `groupBy`, plus the `.toArray.sorted` below, feels painful.

I *think* we can use some `{aggregate,combine}ByKey` trickery like I gave an example of above, as well as some thinking along the lines of my `foldLeft` example even further above, and do all of this in `RDD`-space without having to pull the whole values list into memory.

I'll see if I can get an example together for you.